crates/storage-rocksdb/src/lib.rs

```diff
@@ -275,40 +289,78 @@ impl RocksDBStorage {
         )
     }
 
-    fn prefix_iterator<K: Into<Vec<u8>>>(&self, table: TableKind, prefix: K) -> DBIterator {
+    fn prefix_iterator<K: Into<Vec<u8>>>(
+        &self,
+        table: TableKind,
+        _key_kind: KeyKind,
+        _partition_id: PartitionId,
+        prefix: K,
+    ) -> DBIterator {
         let table = self.table_handle(table);
         let mut opts = ReadOptions::default();
-
+        opts.set_prefix_same_as_start(true);
         opts.set_iterate_range(PrefixRange(prefix));
 
         self.db.raw_iterator_cf_opt(&table, opts)
     }
 
-    fn range_iterator(&self, table: TableKind, range: impl rocksdb::IterateBounds) -> DBIterator {
+    fn range_iterator(
+        &self,
+        table: TableKind,
+        _key: KeyKind,
+        maybe_partition_id: Option<PartitionId>,
+        range: impl rocksdb::IterateBounds,
+    ) -> DBIterator {
         let table = self.table_handle(table);
         let mut opts = ReadOptions::default();
+        if maybe_partition_id.is_none() {
+            // todo: use auto_prefix_mode, at the moment, rocksdb doesn't expose this through the C
+            // binding.
+            opts.set_total_order_seek(true);
+        }
         opts.set_iterate_range(range);
         self.db.raw_iterator_cf_opt(&table, opts)
     }
 
+    #[track_caller]
     fn iterator_from<K: TableKey>(
         &self,
         scan: TableScan<K>,
     ) -> DBRawIteratorWithThreadMode<'_, DB> {
         let scan: PhysicalScan = scan.into();
         match scan {
-            PhysicalScan::Prefix(table, prefix) => {
-                let mut it = self.prefix_iterator(table, prefix.clone());
+            PhysicalScan::Prefix(table, key_kind, partition_id, prefix) => {
+                assert!(table.has_key_kind(&prefix));
+                let mut it = self.prefix_iterator(table, key_kind, partition_id, prefix.clone());
                 it.seek(prefix);
```

Comment:
nit: Should the seek become part of the `*_iterator` methods to fully set up the iterator?